internal/cmd/agent/controller/bundledeployment_controller.go

```diff
@@ -220,6 +231,83 @@ func (r *BundleDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	return ctrl.Result{}, errutil.NewAggregate(merr)
 }
 
+// copyResourcesFromUpstream copies bd's DownstreamResources, from the downstream cluster's namespace on the management
+// cluster to the destination namespace on the downstream cluster, creating that namespace if needed.
+// If bd does not have any DownstreamResources, this method does not issue any API server calls.
+func (r *BundleDeploymentReconciler) copyResourcesFromUpstream(
+	ctx context.Context,
+	bd *fleetv1.BundleDeployment,
+	logger logr.Logger,
+) error {
+	if !experimental.CopyResourcesDownstreamEnabled() {
+		return nil
+	}
+
+	if len(bd.Spec.Options.DownstreamResources) == 0 {
+		return nil
+	}
+
+	destNS := namespaces.GetDeploymentNS(r.DefaultNamespace, bd.Spec.Options)
+
+	ns := corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: destNS}}
+	if err := r.LocalClient.Get(ctx, types.NamespacedName{Name: ns.Name}, &ns); apierrors.IsNotFound(err) {
+		if err := r.LocalClient.Create(ctx, &ns); err != nil {
+			logger.Info(err.Error())
+			return err
+		}
+
+		logger.V(1).Info("Created namespace to copy resources from upstream", "namespace", ns.Name)
+	}
+
+	for _, rsc := range bd.Spec.Options.DownstreamResources {
+		switch strings.ToLower(rsc.Kind) {
+		case "secret":
+			var s corev1.Secret
+			if err := r.Reader.Get(ctx, client.ObjectKey{Namespace: bd.Namespace, Name: rsc.Name}, &s); err != nil {
+				// The bundle deployment is actually created by the bundle reconciler _before_
+				// these objects are copied to the cluster's namespace, hence retries should happen if
+				// they are not found.
+
+				return err
+			}
+
+			s.Namespace = destNS
+			s.ResourceVersion = ""
+			if s.Labels == nil {
+				s.Labels = map[string]string{}
+			}
+			s.Labels["fleet.cattle.io/bundledeployment"] = bd.Name
+
+			if _, err := controllerutil.CreateOrUpdate(ctx, r.LocalClient, &s, func() error { return nil }); err != nil {
```

Comment:
Question: 
Why do we use `CreateOrUpdate` with a _MutateFn_ that just returns `nil`? 

If the object already exists it will do nothing even if there are updates, right? 

`CreateOrUpdate` first calls `Get`, retrieves the object from the cache and if the mutation function does nothing the result is that the object is simply not updated.

Reply:
Great catch, thanks! That's yet more evidence that "if it's not tested, it doesn't work" :þ
I've just updated this implementation to use actual mutate functions, and more broadly support updates to `downstreamResources`. Those resources are of course still not directly monitored, but when copying them downstream, the Fleet agent will now detect updates and force-redeploy the bundle deployment when necessary to take those updates into account.